go/vt/srvtopo/resilient_server.go

```diff
@@ -143,19 +164,30 @@ type srvKeyspaceEntry struct {
 	// the mutex protects any access to this structure (read or write)
 	mutex sync.RWMutex
 
-	// watchRunning describes if the watch go routine is running.
+	// watchState describes if the watch go routine is running.
 	// It is easier to have an explicit field instead of guessing
 	// based on value and lastError.
 	//
-	// if watchrunning is not set, the next time we try to access the
-	// keyspace, we will start a watch.
-	// if watchrunning is set, we are guaranteed to have lastError be
+	// if the state is watchStateIdle, and the time since the last error is
+	// greater than the refresh time, the next time we try to access the
+	// keyspace, we will set watchState to watchStarting and kick off the
+	// watch in a separate goroutine
+	//
+	// in watchStateRunning, we are guaranteed to have lastError be
 	// non-nil and an up-to-date value (which may be nil)
-	watchRunning bool
-	value        *topodatapb.SrvKeyspace
-	lastError    error
+	watchState watchState
 
-	// valueTime is the time when the watch last obtained a non-nil value.
+	// watchStartingCond is used to serialize callers for the first attempt
+	// to establish the watch
+	watchStartingChan chan bool
```

Comment:
nit: these are usually: chan struct{}
and not bool (applies to the other one too)
(ctx.Done() is <-chan struct{})

Reply:
Ah good to know. Done.